nixos/modules/services/torrent/rtorrent.nix

```diff
@@ -0,0 +1,161 @@
+      description = ''
+        If enabled as a system service, the location of rtorrent's config file.
+        If installed as a user service, this value is ignored.
+      '';
+    };
+
+    userDataDir = mkOption {
+      type = types.str;
+      default = "rtorrent";
+      description = ''
+        If installed as a user service, the directory where rtorrent stores its data files relative to $HOME.
+        If enabled as a system service, this value is ignored.
+      '';
+    };
+
+    userConfigFile = mkOption {
+      type = types.str;
+      default = ".rtorrent.rc";
+      description = ''
+        If installed as a user service, the location of rtorrent's config file relative to $HOME.
+        If enabled as a system service, this value is ignored.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "rtorrent";
+      description = ''
+        User account under which rtorrent runs, if enabled as a system service.
+        Ignored if installed as a user service.
+      '';
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "rtorrent";
+      description = ''
+        Group under which rtorrent runs, if enabled as a system service.
+        Ignored if installed as a user service.
+      '';
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.rtorrent;
+      defaultText = "pkgs.rtorrent";
+      description = "The rtorrent package to use.";
+    };
+  };
+
+  config = mkIf (cfg.enable || cfg.install) {
+    users.groups = mkIf (cfg.enable && cfg.group == "rtorrent") {
+      rtorrent.gid = config.ids.gids.rtorrent;
+    };
+    users.extraUsers = mkIf (cfg.enable && cfg.user == "rtorrent") {
+      rtorrent = {
+        uid = config.ids.uids.rtorrent;
+        group = cfg.group;
+        shell = pkgs.bashInteractive;
+        home = cfg.dataDir;
+        createHome = true;
+      };
+    };
+    systemd.services.rtorrent = mkIf cfg.enable {
+      description = "rTorrent system service (via tmux)";
+      after = [ "network.target" ];
+      # Default rtorrent.rc (in preStart) uses bash, which needs to be on the $PATH
+      path = [ cfg.package pkgs.tmux pkgs.bash pkgs.procps ];
+      preStart = ''
+        test -f "${cfg.configFile}" || {
+          echo "creating default rtorrent config file at ${cfg.configFile}."
+          cat > "${cfg.configFile}" << EOF
+      '' + replaceStrings [ "/home/USERNAME/rtorrent" ] [ cfg.dataDir ] (readFile ./rtorrent.rc) + "\nEOF\n}";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "forking";
+        KillMode = "none";
+        User = cfg.user;
+        Group = cfg.group;
+        ExecStart = "${pkgs.tmux}/bin/tmux new-session -c ${cfg.dataDir} -s rtorrent -n rtorrent -d rtorrent -n -o import=${cfg.configFile}";
+        ExecStop = "${pkgs.bash}/bin/bash -c \"tmux send-keys -t rtorrent C-q && while pidof rtorrent > /dev/null; do sleep 0.5; done\"";
```

Comment:
This looks very hacky. Not sure if there is a better way to solve this together with tmux but it doesn't feel right to me.

Reply:
Yeah it's a little reminiscent of creaky init scripts.
On the plus side, it's at least a fairly well tested hack; lots of people have been using it on Arch for a few years:

https://wiki.archlinux.org/index.php/RTorrent#Systemd_services_using_tmux_or_screen